Detectors/AOD/src/AODProducerWorkflowSpec.cxx

```diff
@@ -794,46 +688,272 @@ void AODProducerWorkflowDPL::run(ProcessingContext& pc)
   }
 
-  if (mIgnoreWriter) {
-    std::shared_ptr<arrow::Table> tableCollisions = collisionsBuilder.finalize();
-    std::string tableName("O2collision");
-    writeTableToFile(outfile, tableCollisions, tableName, tfNumber);
-  }
-
   // filling BC table
   // TODO: get real triggerMask
   uint64_t triggerMask = 1;
-  std::sort(BCIDs.begin(), BCIDs.end());
-  uint64_t prevBCid = BCIDs.back();
-  for (auto& id : BCIDs) {
-    if (id == prevBCid) {
-      continue;
-    }
+  for (auto& item : bcsMap) {
+    uint64_t bc = item.first;
     bcCursor(0,
              runNumber,
-             startBCofTF + minGlBC + id,
+             startBCofTF + minGlBC + bc,
              triggerMask);
-    prevBCid = id;
   }
 
-  BCIDs.clear();
-
-  if (mIgnoreWriter) {
-    std::shared_ptr<arrow::Table> tableBC = bcBuilder.finalize();
-    std::string tableName("O2bc");
-    writeTableToFile(outfile, tableBC, tableName, tfNumber);
-  }
+  bcsMap.clear();
 
   // filling mc particles table
   TripletsMap_t toStore;
   fillMCParticlesTable(mcReader, mcParticlesCursor, tracksITSMCTruth, tracksMFTMCTruth, tracksTPCMCTruth, toStore);
-  if (mIgnoreWriter) {
-    std::shared_ptr<arrow::Table> tableMCParticles = mcParticlesBuilder.finalize();
-    std::string tableName("O2mcparticle");
-    writeTableToFile(outfile, tableMCParticles, tableName, tfNumber);
+
+  // collecting tracks for each collisionID
+  std::vector<std::vector<int>> vTracksIdx;
+  int nCollisions = primVertices.size();
+  for (int iCollision = -1; iCollision < nCollisions; iCollision++) {
+    auto prevItem = vCollRefsITS.begin();
+    auto item = std::find(prevItem, vCollRefsITS.end(), iCollision);
+    while (item != vCollRefsITS.end()) {
+      int idx = item - vCollRefsITS.begin();
+      vTracksIdx.push_back({idx, GIndex::Source::ITS});
+      prevItem = item;
+      item = std::find(prevItem + 1, vCollRefsITS.end(), iCollision);
+    }
+    prevItem = vCollRefsMFT.begin();
+    item = std::find(prevItem, vCollRefsMFT.end(), iCollision);
+    while (item != vCollRefsMFT.end()) {
+      int idx = item - vCollRefsMFT.begin();
+      vTracksIdx.push_back({idx, GIndex::Source::MFT});
+      prevItem = item;
+      item = std::find(prevItem + 1, vCollRefsMFT.end(), iCollision);
+    }
+    prevItem = vCollRefsTPC.begin();
+    item = std::find(prevItem, vCollRefsTPC.end(), iCollision);
+    while (item != vCollRefsTPC.end()) {
+      int idx = item - vCollRefsTPC.begin();
+      vTracksIdx.push_back({idx, GIndex::Source::TPC});
+      prevItem = item;
+      item = std::find(prevItem + 1, vCollRefsTPC.end(), iCollision);
+    }
+    prevItem = vCollRefsTPCITS.begin();
+    item = std::find(prevItem, vCollRefsTPCITS.end(), iCollision);
+    while (item != vCollRefsTPCITS.end()) {
+      int idx = item - vCollRefsTPCITS.begin();
+      vTracksIdx.push_back({idx, GIndex::Source::ITSTPC});
+      prevItem = item;
+      item = std::find(prevItem + 1, vCollRefsTPCITS.end(), iCollision);
+    }
+  }
```

Comment:
I guess the abundance of `find` calls will make this Vertex->tracks to tracks->Vertex inversion quite inefficient.
You can see how I do such inversion here: https://github.com/AliceO2Group/AliceO2/blob/e4e46e95ee16c3d6b1a276187daccd11cc22899b/Detectors/Vertexing/src/SVertexer.cxx#L147 (rejecting primary track contributors in https://github.com/AliceO2Group/AliceO2/blob/e4e46e95ee16c3d6b1a276187daccd11cc22899b/Detectors/Vertexing/src/SVertexer.cxx#L154-L156, which you don't need).